mountpoint-s3/src/prefetch/caching_stream.rs

```diff
@@ -229,11 +283,39 @@ where
 
             // Split the body into blocks.
             let mut body: Bytes = body.into();
+            let mut offset = offset;
             while !body.is_empty() {
                 let remaining = (block_size as usize).saturating_sub(self.buffer.len()).min(body.len());
-                let chunk = body.split_to(remaining);
+                let chunk: ChecksummedBytes = body.split_to(remaining).into();
+
+                // We need to return some bytes to the part queue even before we can fill an entire caching block because
+                // we want to start the feedback loop for the flow-control window.
+                //
+                // This is because the read window might be enough to fetch "some data" from S3 but not the entire block.
+                // For example, consider that we got a file system read request with range 2MB to 4MB and we have to start
+                // reading from block_offset=0 and block_size=5MB. The first read window might have a range up to 4MB which
+                // is enough to serve the read request but if the prefetcher is not able to read anything it cannot tell
+                // the stream to move its read window.
+                //
+                // A side effect from this is delay on the cache updating which is actually good for performance but it makes
+                // testing a bit more complicated because the cache might not be updated immediately.
+                let part_range = self
```

Comment:
Why the duplication with `make_part`? I'm struggling to see what is different and why. Could we review this logic and maybe extract or document it?

Reply:
`make_part` only trim data to the block boundaries but here we want to push some data (i.e. `chunk`) to the part queue even before we have a full block. The reasons are in above comments. I will extract it to a new function.